### ISSUE TITLE
Enum variants can't be renamed

### ISSUE THREAD

**Issue by sfackler** [serde-rs/serde#182]
I would expect this to work, but it doesn't appear to:

```rust
#[derive(Deserialize)]
enum Foo {
    #[serde(rename = "BAR")]
    Bar,
}
```

**Comment by jimmycuadra** [comment]
#195 sounds like the same thing.

**Comment by JohnHeitmann** [comment]
I started work on this and have a [decent implementation](https://github.com/JohnHeitmann/serde/tree/enum-rename), but there are a couple blockers to release.

First, serde previously allowed the rename attribute to be applied to enum variants and would silently do nothing. If my change is released as-is where it re-uses the sensible attribute naming scheme already in place, it could silently break folks in really surprising ways. I'm going to work on attribute validation, but even if attribute validation goes out before enum rename support, it's not a guarantee that we catch everyone with the problem. So, the options are:

1. Release with the same names and accept that we could silently break folks
2. Release with the same names, but only after validation has been deployed to crates.io for at least a full major release cycle (which is still not a guarantee)
3. Release with a different name like "rename-variant" and leave a permanent naming wart

#3 is unfortunately probably the winner.

Second, format-specific renaming is a little bit broken. I think it's a problem that already affects field renames, so it might not be a hard blocker, but it has me concerned.

**Comment by sfackler** [comment]
I think option 1 might be reasonable, honestly. I can't imagine there are many cases where people wanted to rename an enum but then didn't realize that it didn't work. Could be wrong though!

**Comment by jimmycuadra** [comment]
Option 1 also seems like the best choice to me. Not to mention serde is < 1.0 so backwards compatibility is not a hard requirement according to SemVer.

**Comment by erickt** [comment]
I think (1) is reasonable. This will go through a major version bump, and there will be some breakage anyway. We'll just have to document this change.

**Comment by erickt** [comment]
I just swept through all the github usages of `#[serde(rename=...)]`. I couldn't find anyone using it on an enum.

**Comment by erickt** [comment]
This was implemented in #233 for version v0.7.0.
